akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -137,6 +139,203 @@ import akka.util.ccompat.JavaConverters._
+    override def tryPull(): Unit = ()
+    override def cancel(): Unit = ()
+    override def isClosed: Boolean = true
+  }
+}
+
+@InternalApi
+private[akka] final class FlattenConcat[T, M](parallelism: Int)
+    extends GraphStage[FlowShape[Graph[SourceShape[T], M], T]] {
+  require(parallelism >= 1, "parallelism should >= 1")
+  private val in = Inlet[Graph[SourceShape[T], M]]("flattenConcat.in")
+  private val out = Outlet[T]("flattenConcat.out")
+
+  override def initialAttributes: Attributes = DefaultAttributes.flattenConcat
+  override val shape: FlowShape[Graph[SourceShape[T], M], T] = FlowShape(in, out)
+  override def createLogic(inheritedAttributes: Attributes) =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      import FlattenConcat._
+      private var inflightSources: BufferImpl[InflightSource[T]] = _
+
+      override def preStart(): Unit = inflightSources = BufferImpl(parallelism, inheritedAttributes)
+
+      override def onPush(): Unit = {
+        val source = grab(in)
+        addSource(source)
+        if (!inflightSources.isFull) {
+          tryPull(in)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = if (inflightSources.isEmpty) completeStage()
+
+      override def onPull(): Unit = {
+        pull(in)
+        setHandler(
+          out,
+          new OutHandler {
+            override def onPull(): Unit = {
+              if (inflightSources.nonEmpty) {
+                val currentSource = inflightSources.peek()
+                //purge if possible
+                if (currentSource.hasNext) {
+                  push(out, currentSource.next)
+                  if (currentSource.isClosed) {
+                    removeSource(currentSource)
+                  }
+                } else if (currentSource.isClosed) {
+                  removeSource(currentSource)
+                } else {
+                  currentSource.tryPull()
+                }
+              } else if (!hasBeenPulled(in))
+                tryPull(in)
+            }
+          })
+      }
+
+      private def addSource(source: Graph[SourceShape[T], M]): Unit = {
+        TraversalBuilder.getValuePresentedSource(source) match {
+          case OptionVal.Some(graph) =>
+            graph match {
+              case single: SingleSource[T] @unchecked =>
+                if (isAvailable(out) && inflightSources.isEmpty) {
+                  push(out, single.elem)
+                } else {
+                  inflightSources.enqueue(new InflightSingleSource(single.elem))
+                }
+              case iterable: IterableSource[T] @unchecked =>
+                val inflightSource = new InflightIterableSource[T](iterable.elements)
+                if (isAvailable(out) && inflightSources.isEmpty) {
+                  if (inflightSource.hasNext) {
+                    push(out, inflightSource.next)
+                    if (inflightSource.hasNext) {
+                      inflightSources.enqueue(inflightSource)
```

Comment:
Do we have optimizations for IterableSource in other places? I wonder if that might be overkill here? I'm not sure I follow exactly, we push the first element from the IterableSource but then add the source to the queue. No need for `attachAndMaterializeSource` in that case, or is that handled later?

Might be enough with the optimization for the SingleSource in same way as we have in FlattenMerge?

Reply:
No, I was want to optimize it but not complete the PR. and the benchmark do show some nearly 10x performance booster.

Reply:
I would prefer not adding that optimization in this PR but adding that separately and focusing on the added operator/api here.

Reply:
Ok, I see now why you have combined that in here, might be ok.

Reply:
I can seperate it

Reply:
Retracted that request somewhat after reviewing a bit more, I think you can leave it in